[build] Consolidation of build system infrastructure (#15560)
coq/coq: dev/doc/build-system.dune.md

unless you are adding a new binary, library, or plugin, or want to
tweak some low-level option.

## The bootstrap process / rule generation

> todo

Reviewer: This needs to be finished

unless you are adding a new binary, library, or plugin, or want to
tweak some low-level option.

## The bootstrap process / rule generation

Dune is able to build all the OCaml parts of Coq in a pretty standard

> way, using its built-in rule generation for OCaml. Public tTools

Reviewer: ```suggestion
way, using its built-in rule generation for OCaml. Public tools
```

written in OCaml are distributed in the `coq-core` package.

> The set of public `.v` files present in this repos, usually referred

Reviewer: ```suggestion
The set of public `.v` files present in this repo, usually referred
```

as the "Coq Standard Library" are distributed in the `coq-stdlib`

> package. As of June 2022, Dune has to a set of built-in rules for `.v`

Reviewer: ```suggestion
package. As of June 2022, Dune has a set of built-in rules for `.v`
```

> which is capable of building Coq's standard library fine.

Reviewer: ```suggestion
files which is capable of building Coq's standard library.
```

However, in order to have a bit more control, we generate ourselves a
set of custom rules using the `tools/dune_rule_gen` binary, which are
then stored in the `theories/dune` file. This allows us to have a
finer control over the build rules without having to bump the Dune
version. The generation of the `theories/dune` and
`user-contrib/*/dune` files is known as "bootstrap".

The rule generation program in `tools/dune_rule_gen` is mostly derived
from Dune's built-in rules, and it works in an straightforward way: it

> will scan a directory with `.v` files on it, and output the

Reviewer: ```suggestion
will scan a directory with `.v` files in it, and output the
```

corresponding build rule. The script will look at some configuration
values such as whether native is enabled or not and adapt rule
generation accordingly.

In the case of native, the script supports two modes, `coqc
-native-compiler on` and `coqnative`. The default is the first, as

> currently `coqnative` incurs in a 33% build time overhead in a pretty

Reviewer: ```suggestion
currently `coqnative` incurs a 33% build time overhead even on a pretty
```

## OPAM file generation

`.opam` files will be automatically generated by Dune from the package
descriptions in the `dune-project` file; see Dune's manual for more

> details. For now we have disabled this due to some bugs.

Reviewer: What is the exact reason we disabled this?

Author: Dune bugs, in particular `dune subst` was broken for source tree with unicode files. We shall re-enable soon, hopefully.

Reviewer: This has been fixed, lets generate opam again.

Author: That requires a dune version bump which may have other implications, I suggest we do that in a different PR, this PR is large enough as it is.

Reviewer: I've created an issue for it https://github.com/coq/coq/issues/16071
